merkle-tree/bounded-vec/src/lib.rs

```diff
@@ -520,31 +522,293 @@ where
 
 impl<'a, T> Eq for CyclicBoundedVec<'a, T> where T: Clone + Eq + Pod {}
 
+pub struct CyclicBoundedVecIterator<'a, T>
+where
+    T: Clone + Pod,
+{
+    vec: &'a CyclicBoundedVec<'a, T>,
+    current: usize,
+    is_finished: bool,
+}
+
+impl<'a, T> Iterator for CyclicBoundedVecIterator<'a, T>
+where
+    T: Clone + Pod,
+{
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.is_finished {
+            None
+        } else {
+            if self.current == self.vec.last_index {
+                self.is_finished = true;
+            }
+            let new_current = (self.current + 1) % self.vec.capacity();
+            let element = self.vec.get(self.current);
+            self.current = new_current;
+            element
+        }
+    }
```

Comment:
how does this achieve equivalence to what it is meant to replace?
Can you still iterate over the cyclic vec in a normal way?

Reply:
Depends what you mean by "normal way". I was hoping that the tests make it clear, but the iterator basically iterates from the oldest (not necessarily `0`!) element to the newest one.

Reply:
If your question is more about "how the heck all this code creates an iterator" - it was confusing for me as well some time ago. These resources were helpful

https://dev.to/wrongbyte/implementing-iterator-and-intoiterator-in-rust-3nio

Reply:
Also

https://doc.rust-lang.org/std/iter/trait.Iterator.html
https://refactoring.guru/design-patterns/iterator/rust/example#example-1